src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPartDiscoveryModel.cs

```diff
@@ -0,0 +1,188 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationParts
+{
+    public class AssemblyPartDiscoveryModel : IEquatable<AssemblyPartDiscoveryModel>
+    {
+        internal static readonly string[] ViewsAssemblySuffixes = new[]
+        {
+            ".PrecompiledViews",
+            ".Views",
+        };
+
+        private readonly List<AssemblyPartDiscoveryModel> _additionalParts;
+
+        public AssemblyPartDiscoveryModel(Assembly assembly)
+            : this(assembly, assembly.GetName().Name, assembly.GetCustomAttributes(inherit: false))
+        {
+        }
+
+        public AssemblyPartDiscoveryModel(
+            Assembly assembly, 
+            string name,
+            IReadOnlyList<object> attributes)
+        {
+            Assembly = assembly;
+            Name = name;
+            Attributes = attributes;
+            _additionalParts = new List<AssemblyPartDiscoveryModel>();
+        }
+
+        public string Name { get; }
+
+        public Assembly Assembly { get; }
+
+        public IReadOnlyList<object> Attributes { get; }
+
+        public AssemblyPartDiscoveryModel Parent { get; private set; }
+
+        public IReadOnlyList<AssemblyPartDiscoveryModel> AdditionalParts => _additionalParts;
```

Comment:
So are these a tree? This really makes it seem like it is (both parent and children)